### ISSUE TITLE
Using <input> inside a contenteditable="false" is broken in Firefox

### ISSUE THREAD

**Issue by Zhouzi** [ianstormtaylor/slate#1744]
We noticed a bug in Firefox, rendering a node as follows is broken:

```html
<div contenteditable="false">
  <input type="text" />
</div>
```

There are two things to note:

1. Left/right arrow keys don't move the caret
2. Pressing a character key (or backspace) throws an error:

> Permission denied to access property "nodeType"

I created a [repository](https://github.com/Zhouzi/slate-firefox-bug) reproducing the error with Slate and a minimal setup:
http://gabinaureche.com/slate-firefox-bug/

The error comes from `isInEditor` which tries to read `window.getSelection().anchorNode.nodeType` but `anchorNode` is "Restricted".

This bug is reproducible without Slate, here's a pretty simple scenario:
https://jsfiddle.net/hsqebsk2/1/

I filed an issue to Firefox:
https://bugzilla.mozilla.org/show_bug.cgi?id=1449936

Probably related to #843

**Comment by CameronAckermanSEL** [comment]
This can be worked around using 
```
<div contenteditable>
    <div contenteditable="false">
        <input contenteditable="true" type="text" value="Some text content">
    </div>
</div>
```

But if this bug is reproducible outside of slate, I think this is an indication that this is not a slate bug.

**Comment by CameronAckermanSEL** [comment]
This can be worked around by using 

```

<div contenteditable>
    <div contenteditable="false">
        <input type="text" value="Some text content">
    </div>
</div>
```

Or by using the componentDidThrow() react event handler to keep these errors from crashing the editor.

I also don't think this is a Slate bug.

**Comment by Zhouzi** [comment]
Thanks for the suggestion @CameronAckermanSEL 👍 

I gave it a try and it does work outside of Slate: https://jsfiddle.net/hsqebsk2/4/

But doesn't in the context of Slate 😕

**Comment by nkoder** [comment]
not sure whether it will help anyone, but I have similar issue, and here is my code sample to reproduce it on Firefox: https://codepen.io/nkoder/pen/pKLQWz

**Comment by Iuriy-Budnikov** [comment]
I have the same issue +1

**Comment by Soreine** [comment]
Either we try to make Slate work around the bug, or we find a way to pressure Firefox into conforming to other browser's behavior. We have **[filed this bug to Mozilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1449936)**, but it is not getting much attention...

**Comment by ianstormtaylor** [comment]
I'm open to not having this be the structure we use in Slate, I think the tricky part is finding a DOM structure for voids that satisfies the many different bugs/behaviors in browsers. It's hard to keep track of all of the pitfalls, which is why we need more inline comments/docs for this kind of thing probably. There was probably a good reason it was switched to this (now problematic) behavior in the first place.

**Comment by Soreine** [comment]
Thanks to #2178 , Firefox will no longer crashes. However, the inputs are nearly unusable, because you cannot move the cursor freely inside it, even by clicking somewhere.

Here is another error thrown when focusing the input:

```
Error: Permission denied to access property "window" index.js:25
isWindow                     index.js:25
getWindow                    index.js:39
findRange                    slate-react.es.js:1409
onEvent                      slate-react.es.js:2029
Content/</_this[handler]     slate-react.es.js:1936
```

**Comment by CameronAckermanSEL** [comment]
I imagine this might sound kinda dumb, but maybe the thing to do is to just use a simple inline or block that is styled to look like an input as a work around? I get that this is really frustrating and might get people by until mozilla addresses this issue.

**Comment by Iuriy-Budnikov** [comment]
Any updates?

**Comment by dmitrizzle** [comment]
Hi @ianstormtaylor, it seems that this issue and #1926 are the same, perhaps we could merge them?

I'm also hoping that participants here would be willing to test the `embeds` example in other browsers and confirm broken behaviour. Would be more than happy to contribute if I knew where to start; this is a blocker for my project and, presumably, any project that has `<input />` elements inside void blocks with the newer Slate releases.

**Comment by Iuriy-Budnikov** [comment]
I'll be able to help as well.

**Comment by skogsmaskin** [comment]
I think this might be related https://github.com/ianstormtaylor/slate/issues/2116

**Comment by gnestor** [comment]
Related to https://github.com/ianstormtaylor/slate/issues/191 too

**Comment by gnestor** [comment]
In https://github.com/ianstormtaylor/slate/issues/2261, @klis87 points out that `onChange` events were firing for inputs in a void node in slate 0.41 and now they aren't in 0.42. This may be a separate issue than the Firefox one being discussed here as it affects all browsers as far as I can tell.

I'd be happy to hunt this regression down but where should I start looking?

**Comment by gnestor** [comment]
I was able to hunt this down in https://github.com/ianstormtaylor/slate/pull/2198.

Simply reverting this commit fixes the embed example and allows elements such as input elements within a void node to receives change, keydown, click, etc. events.

#2198 was intended to make copying behavior consistent for inline void with text and inline void without text. However, it mistakes void blocks as being `inTheEditor`.

@uipoptart Can this be implemented differently?

**Comment by uipoptart** [comment]
Hey @gnestor, sorry that change is causing trouble for you. Your use case is another reason why #2072 would be great for the entire community. We all have complicated use cases and it is hard to keep them from colliding. Short of implementing #2072 we could add a single prop to the inline allowing it to opt in/out of being considered `inTheEditor` or maybe we could look to see if any children are their own inputs who'd want to be considered first. 
Currently, I don't have time approved to work on writing a commit to bridge both behaviors but I am happy to help brainstorm solutions that could solve both our problems. Thanks for tagging me!

**Comment by ianstormtaylor** [comment]
I've rolled back the regression that made inputs inside voids unable to be focused in all browsers, so this issue can remain for the Firefox case.

**Comment by CodePlayer** [comment]
I have the same issue +1.
I'm a new comer. However, the slate is still considered buggy.

**Comment by Zhouzi** [comment]
The [bug report](https://bugzilla.mozilla.org/show_bug.cgi?id=1449936) on Firefox's side was just marked as resolved. I wasn't able to reproduce the bug in the two following scenarios:

* https://www.slatejs.org/#/embeds
* https://jsfiddle.net/hsqebsk2/

So it does seem to be fixed. Could someone else confirm?

**Comment by josephluck** [comment]
> So it does seem to be fixed. Could someone else confirm?

It seems a little better, however if you edit from within the youtube link (i.e not at the end of the link) the cursor jumps to the end.

**Comment by CameronAckermanSEL** [comment]
I think this got fixed in a recent PR. Going to close

**Comment by dimileeh** [comment]
I would like to comment for someone who would be looking for a solution around a similar issue.
I installed `slate` and other related `slate` packages via `yarn`. The latest version as of the moment of writing was `0.57.1` as is shown here: https://www.npmjs.com/package/slate-react.

No matter what I tried, `input` element inside Slate simply wouldn't work. Yes, I could focus and put the cursor at any place within the `input`, but typing anything produced nothing. Pressing Backspace inside the `input` field would delete the whole block within Slate. I've even downloaded the source code for the `editable-voids` example from https://www.slatejs.org/examples/editable-voids. Still, it just didn't work in my Gatsby app.

So I cloned the slate repository and added the slate packages located in the `packages` to my local project. Cleared the cache (by deleting the `node_modules` folder), installed the dependencies again (`yarn install`).

After that, I rebuilt my Gatsby project – now with Slate dependencies fetched from my local folder, cloned from the slate repo as of the latest revision `f6bfe034d707693488c38da77537fd36cb8856cf`.

And it worked! I can now type inside input fields within Slate! Spent 6 hours trying to debug this issue.

I suppose the issue was fixed in #3516 but the version 0.57.2 hasn't been published yet.
